### PR TITLE
[io-layer] Remove FormatMessage

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by vargaz** [comment]
We need to add linker support for this, the runtime code had #ifndef DISABLE_FULL_MESSAGES, the new code doesn't.

File: mcs/class/System/ReferenceSources/Win32Exception.cs [hunk 1]
```diff
@@ -1,15 +1,1752 @@
...
 namespace System.ComponentModel
 {
 	partial class Win32Exception
 	{
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		internal static extern string W32ErrorMessage (int error_code);
+		const int FORMAT_MESSAGE_IGNORE_INSERTS = 0x00000200;
+		const int FORMAT_MESSAGE_FROM_SYSTEM    = 0x00001000;
+
+		[DllImport ("api-ms-win-core-localization-l1-2-0.dll", CharSet = CharSet.Unicode,
```
Comment (marek-safar) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: This is wrong, it's CoreClr library which we never install in-place

File: mcs/class/System/ReferenceSources/Win32Exception.cs [hunk 2]
```diff
@@ -1,15 +1,1752 @@
...
+			}
+
+			string message;
+			if (!Messages.Value.TryGetValue (error, out message))
+				return string.Format ("mono-io-layer-error ({0})", error);
+
+			return message;
+		}
+
+		static Lazy<Dictionary<int, string>> Messages =
```
Comment (marek-safar) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: use simple C# switch, it does lazy init for you checks for duplicate values, etc

File: mcs/class/System/ReferenceSources/Win32Exception.cs [hunk 3]
```diff
@@ -1,15 +1,1752 @@
...
+			new Lazy<Dictionary<int, string>> (BuildMessages, LazyThreadSafetyMode.ExecutionAndPublication);
 
-		private static string GetErrorMessage (int error)
+		static Dictionary<int, string> BuildMessages ()
 		{
-			return W32ErrorMessage (error);
+			return new Dictionary<int, string> () {
```
Comment (marek-safar) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: For MOBILE you need to use only most common one and for the rest print message with the error code

File: mcs/class/System/ReferenceSources/Win32Native.cs [hunk 4]
```diff
@@ -0,0 +1,1790 @@
+using System.IO;
+
+namespace Microsoft.Win32
+{
+	static class Win32Native
```
Comment (marek-safar) [inline mcs/class/System/ReferenceSources/Win32Native.cs]: It looks like the constants can be local to the method with switch and save another huge class with fields space

File: mcs/class/System/ReferenceSources/Win32Exception.cs (hunk 2 above)
Comment (ludovic-henry) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: how about the performance? switch has complexity O(n) while the dictionary has complexity O(1). And creating this dictionary takes 3ms on my machine.

Comment (evincarofautumn) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: Wait, do we always produce O(n) code for `switch`? It can be O(1) or O(log n) in most cases.

Comment (vargaz) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: I think mcs converts a switch on strings to a dictionary behind the scene.

Comment (marek-safar) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: Correct, switch is optimized not sure why you think it's O(n) in this case

Comment (ludovic-henry) [inline mcs/class/System/ReferenceSources/Win32Exception.cs]: @vargaz @marek-safar ok, good to know, I didn't know we had this optimisation

**Comment by ludovic-henry** [comment]
@marek-safar I updated with a switch and extended the `DISABLE_FULL_MESSAGES` to System.

File: configure.ac [hunk 5]
```diff
@@ -4419,6 +4419,10 @@ fi
       fi
     fi
 
+    if test "x$mono_feature_disable_full_messages" != "xyes"; then
+      echo "HAVE_FULL_MESSAGES=1" >> $srcdir/$mcsdir/build/config.make
```
Comment (marek-safar) [inline configure.ac]: I don't like this. I think it's overkill and it should be build profile driven for example we can completely ifdef it out for MOBILE

Comment (vargaz) [inline configure.ac]: I agree we should just ifdef it out of MOBILE.

Comment (ludovic-henry) [inline configure.ac]: It make things easier, I will change it to MOBILE.

**Comment by ludovic-henry** [comment]
@mareks I updated with `MOBILE`

**Pull request merged**